From ec99adf91734ab9b301a8ed731cca84de0422910 Mon Sep 17 00:00:00 2001
From: "Ronald S. Bultje" <rsbultje@gmail.com>
Date: Wed, 27 Aug 2025 09:40:39 -0400
Subject: [PATCH 2/4] frame header debugging

---
 aom/debug.h               |   1 +
 av1/decoder/decodeframe.c | 179 +++++++++++++++++++++++++++++++++++++-
 2 files changed, 179 insertions(+), 1 deletion(-)

diff --git a/aom/debug.h b/aom/debug.h
index 3339b3439f..8c8847f1eb 100644
--- a/aom/debug.h
+++ b/aom/debug.h
@@ -30,5 +30,6 @@
 
 #define DEBUG_OBU_HDR 0
 #define DEBUG_SEQ_HDR 0
+#define DEBUG_FRAME_HDR 0
 
 #endif /* AOM_DEBUG_H */
diff --git a/av1/decoder/decodeframe.c b/av1/decoder/decodeframe.c
index 6a04bf6a0e..70e22ec4ea 100644
--- a/av1/decoder/decodeframe.c
+++ b/av1/decoder/decodeframe.c
@@ -6333,6 +6333,11 @@ static AOM_INLINE void read_film_grain(AV1_COMMON *cm,
                                        struct aom_read_bit_buffer *rb) {
   if (cm->seq_params.film_grain_params_present) {
     av1_read_film_grain_params(cm, rb);
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-filmgrain[%d]: off=%u\n",
+           cm->cur_frame->film_grain_params_present,
+           rb->bit_offset);
+#endif
   } else {
     memset(&cm->film_grain_params, 0, sizeof(cm->film_grain_params));
   }
@@ -8303,6 +8308,11 @@ static int read_uncompressed_header(AV1Decoder *pbi,
     }
 #endif  // CONFIG_MULTI_FRAME_HEADER
 #if CONFIG_CWG_F317
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-ids[f=%d,s=%d]: off=%u\n",
+           cm->cur_mfh_id, cm->mfh_params[cm->cur_mfh_id].mfh_seq_header_id,
+           rb->bit_offset);
+#endif
   }
 #endif  // CONFIG_CWG_F317
 
@@ -8342,6 +8352,10 @@ static int read_uncompressed_header(AV1Decoder *pbi,
     }
 #endif  // CONFIG_CWG_F317
     pbi->reset_decoder_state = 0;
+#if DEBUG_FRAME_HDR
+      printf("HDR: post-show_existing_frame[%d]: off=%u\n",
+             cm->show_existing_frame, rb->bit_offset);
+#endif
 
     if (cm->show_existing_frame) {
       if (pbi->sequence_header_changed) {
@@ -8418,6 +8432,10 @@ static int read_uncompressed_header(AV1Decoder *pbi,
       } else {
         current_frame->refresh_frame_flags = 0;
       }
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-existing_frame_idx[%d]: off=%u\n",
+           existing_frame_idx, rb->bit_offset);
+#endif
 
       return 0;
     }
@@ -8460,17 +8478,29 @@ static int read_uncompressed_header(AV1Decoder *pbi,
 #if CONFIG_CWG_F317
     }
 #endif  // CONFIG_CWG_F317
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-frametype[%d]: off=%u\n",
+           current_frame->frame_type, rb->bit_offset);
+#endif
 #if CONFIG_RANDOM_ACCESS_SWITCH_FRAME
     current_frame->long_term_id = -1;
     if (current_frame->frame_type == KEY_FRAME) {
       current_frame->long_term_id =
           aom_rb_read_literal(rb, seq_params->number_of_bits_for_lt_frame_id);
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-ltr_id[%d]: off=%u\n",
+           current_frame->long_term_id, rb->bit_offset);
+#endif
     } else if (obu_type == OBU_RAS_FRAME) {
       cm->num_ref_key_frames = aom_rb_read_literal(rb, 3);
       for (int i = 0; i < cm->num_ref_key_frames; i++) {
         cm->ref_long_term_ids[i] =
             aom_rb_read_literal(rb, seq_params->number_of_bits_for_lt_frame_id);
       }
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-ltr_setup[%d]: off=%u\n",
+           cm->num_ref_key_frames, rb->bit_offset);
+#endif
     }
 #endif  // CONFIG_RANDOM_ACCESS_SWITCH_FRAME
     if (pbi->sequence_header_changed) {
@@ -8570,6 +8600,11 @@ static int read_uncompressed_header(AV1Decoder *pbi,
 #endif  // CONFIG_CWG_F317
 #endif  // !CONFIG_F322_OBUER_ERM
   }
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-frametype_bits[%d,show=%d|%d]: off=%u\n",
+           current_frame->frame_type, cm->show_frame, cm->showable_frame,
+           rb->bit_offset);
+#endif
 
   av1_set_frame_sb_size(cm, cm->seq_params.sb_size);
 
@@ -8629,6 +8664,11 @@ static int read_uncompressed_header(AV1Decoder *pbi,
 #endif  // CONFIG_CWG_F317
     }
   }
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-frame_size_override_flag[%d,poc=%d,p_ref=%d]: off=%u\n",
+           frame_size_override_flag, current_frame->order_hint,
+           pbi->signal_primary_ref_frame, rb->bit_offset);
+#endif
 
 #if !CONFIG_CWG_F293_BUFFER_REMOVAL_TIMING
   if (seq_params->decoder_model_info_present_flag) {
@@ -8661,6 +8701,10 @@ static int read_uncompressed_header(AV1Decoder *pbi,
         }
       }
     }
+#if DEBUG_FRAME_HDR
+        printf("HDR: post-buffer_removal_timings[%d]: off=%u\n",
+               cm->buffer_removal_time_present, rb->bit_offset);
+#endif
   }
 #endif  // CONFIG_CWG_F293_BUFFER_REMOVAL_TIMING
   if (current_frame->frame_type == KEY_FRAME) {
@@ -8778,6 +8822,10 @@ static int read_uncompressed_header(AV1Decoder *pbi,
       }
     }
   }
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-refresh_frame_flags[%x]: off=%u\n",
+           current_frame->refresh_frame_flags, rb->bit_offset);
+#endif
 
 #if CONFIG_F322_OBUER_ERM
 #if CONFIG_RANDOM_ACCESS_SWITCH_FRAME
@@ -8884,6 +8932,10 @@ static int read_uncompressed_header(AV1Decoder *pbi,
 
     features->tip_frame_mode = TIP_FRAME_DISABLED;
     setup_frame_size(cm, frame_size_override_flag, rb);
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-framesize[%dx%d]: off=%u\n",
+           cm->width, cm->height, rb->bit_offset);
+#endif
     read_screen_content_params(cm, rb);
     features->allow_intrabc = aom_rb_read_bit(rb);
     if (features->allow_intrabc) {
@@ -8897,6 +8949,16 @@ static int read_uncompressed_header(AV1Decoder *pbi,
     cm->prev_frame = NULL;
 
     cm->cur_frame->num_ref_frames = 0;
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-screencontent_ibc[%d,%d,%d,%d,%d,%d]: off=%u\n",
+           features->allow_screen_content_tools,
+           features->cur_frame_force_integer_mv,
+           features->allow_intrabc,
+           features->allow_global_intrabc,
+           features->allow_local_intrabc,
+           features->max_bvp_drl_bits,
+           rb->bit_offset);
+#endif
   } else {
     cm->current_frame.temporal_layer_id = cm->tlayer_id;
     cm->current_frame.mlayer_id = cm->mlayer_id;
@@ -8907,6 +8969,10 @@ static int read_uncompressed_header(AV1Decoder *pbi,
       cm->cur_frame->film_grain_params_present =
           seq_params->film_grain_params_present;
       setup_frame_size(cm, frame_size_override_flag, rb);
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-framesize[%dx%d]: off=%u\n",
+           cm->width, cm->height, rb->bit_offset);
+#endif
       read_screen_content_params(cm, rb);
       features->allow_intrabc = aom_rb_read_bit(rb);
       if (features->allow_intrabc) {
@@ -8917,6 +8983,16 @@ static int read_uncompressed_header(AV1Decoder *pbi,
       }
 
       cm->cur_frame->num_ref_frames = 0;
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-screencontent_ibc[%d,%d,%d,%d,%d,%d]: off=%u\n",
+           features->allow_screen_content_tools,
+           features->cur_frame_force_integer_mv,
+           features->allow_intrabc,
+           features->allow_global_intrabc,
+           features->allow_local_intrabc,
+           features->max_bvp_drl_bits,
+           rb->bit_offset);
+#endif
 
     } else if (pbi->need_resync != 1) { /* Skip if need resync */
       // Implicitly derive the reference mapping
@@ -9332,6 +9408,16 @@ static int read_uncompressed_header(AV1Decoder *pbi,
         if (features->allow_intrabc) {
           read_frame_max_bvp_drl_bits(cm, rb);
         }
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-screencontent_ibc[%d,%d,%d,%d,%d,%d]: off=%u\n",
+           features->allow_screen_content_tools,
+           features->cur_frame_force_integer_mv,
+           features->allow_intrabc,
+           features->allow_global_intrabc,
+           features->allow_local_intrabc,
+           features->max_bvp_drl_bits,
+           rb->bit_offset);
+#endif
 
         if (features->cur_frame_force_integer_mv) {
           features->fr_mv_precision = MV_PRECISION_ONE_PEL;
@@ -9432,6 +9518,12 @@ static int read_uncompressed_header(AV1Decoder *pbi,
                                "Reference frame has invalid dimensions");
         }
       }
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-frametype-specific-bits[mvprec:%d,flt:%d,mm:%d]: off=%u\n",
+           features->fr_mv_precision - 3,
+           features->interp_filter, features->enabled_motion_modes,
+           rb->bit_offset);
+#endif
     }
   }
 
@@ -9641,8 +9733,18 @@ static int read_uncompressed_header(AV1Decoder *pbi,
 #if CONFIG_CWG_F317
   }
 #endif  // CONFIG_CWG_F317
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-refresh_context[%d,%d]: off=%u\n",
+           features->disable_cdf_update, features->refresh_frame_context,
+           rb->bit_offset);
+#endif
 
   read_tile_info(pbi, rb);
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-tiling[%dx%dtiles,%dbytes]: off=%u\n",
+           pbi->common.tiles.cols, pbi->common.tiles.rows, pbi->tile_size_bytes,
+           rb->bit_offset);
+#endif
   if (!av1_is_min_tile_width_satisfied(cm)) {
     aom_internal_error(&cm->error, AOM_CODEC_CORRUPT_FRAME,
                        "Minimum tile width requirement not satisfied");
@@ -9654,6 +9756,16 @@ static int read_uncompressed_header(AV1Decoder *pbi,
   cm->cur_frame->u_ac_delta_q = quant_params->u_ac_delta_q;
   cm->cur_frame->v_ac_delta_q = quant_params->v_ac_delta_q;
   xd->bd = (int)seq_params->bit_depth;
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-quant[yac:%d,deltas=ydc:%d,uac:%d/dc:%d,vac:%d/dc:%d]: off=%u\n",
+           quant_params->base_qindex,
+           quant_params->y_dc_delta_q,
+           quant_params->u_ac_delta_q,
+           quant_params->u_dc_delta_q,
+           quant_params->v_ac_delta_q,
+           quant_params->v_dc_delta_q,
+           rb->bit_offset);
+#endif
 
   set_primary_ref_frame_and_ctx(pbi);
 
@@ -9671,12 +9783,22 @@ static int read_uncompressed_header(AV1Decoder *pbi,
   }
 
   setup_segmentation(cm, rb);
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-segmentation[%d]: off=%u\n",
+           cm->seg.enabled,
+           rb->bit_offset);
+#endif
 
   setup_qm_params(&cm->seq_params,
 #if CONFIG_CWG_E242_SEQ_HDR_ID
                   pbi->active_seq,
 #endif  // CONFIG_CWG_E242_SEQ_HDR_ID
                   quant_params, cm->seg.enabled, av1_num_planes(cm), rb);
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-qm[%d]: off=%u\n",
+           quant_params->using_qmatrix,
+           rb->bit_offset);
+#endif
 
   cm->delta_q_info.delta_q_res = 1;
   cm->delta_q_info.delta_lf_res = 1;
@@ -9694,6 +9816,12 @@ static int read_uncompressed_header(AV1Decoder *pbi,
       av1_reset_loop_filter_delta(xd, av1_num_planes(cm));
     }
   }
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-delta_q_lf[q:%d,lf:%d]: off=%u\n",
+           cm->delta_q_info.delta_q_present_flag,
+           cm->delta_q_info.delta_lf_present_flag,
+           rb->bit_offset);
+#endif
 
   xd->cur_frame_force_integer_mv = features->cur_frame_force_integer_mv;
 #if CONFIG_DISABLE_LOOP_FILTERS_LOSSLESS
@@ -9754,6 +9882,11 @@ static int read_uncompressed_header(AV1Decoder *pbi,
     features->allow_parity_hiding = false;
   else
     features->allow_parity_hiding = aom_rb_read_bit(rb);
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-tcq_parity[tcq:%d,par:%d]: off=%u\n",
+           features->tcq_mode, features->allow_parity_hiding,
+           rb->bit_offset);
+#endif
 
   setup_segmentation_dequant(cm, xd);
 
@@ -9773,23 +9906,51 @@ static int read_uncompressed_header(AV1Decoder *pbi,
     cm->rst_info[2].frame_filters_on = 0;
   }
   setup_loopfilter(cm, rb);
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-deblock[y:%d|%d,u:%d,v:%d]: off=%u\n",
+           cm->lf.filter_level[0], cm->lf.filter_level[1],
+           cm->lf.filter_level_u, cm->lf.filter_level_v,
+           rb->bit_offset);
+#endif
 
   if (!features->coded_lossless && seq_params->enable_gdf) {
     setup_gdf(cm, rb);
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-gdf[%d]: off=%u\n",
+           cm->gdf_info.gdf_mode,
+           rb->bit_offset);
+#endif
   } else {
     cm->gdf_info.gdf_mode = 0;
   }
   if (!features->coded_lossless && seq_params->enable_cdef) {
     setup_cdef(cm, rb);
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-cdef[%d]: off=%u\n",
+           cm->cdef_info.cdef_frame_enable,
+           rb->bit_offset);
+#endif
   }
   if (!features->all_lossless && seq_params->enable_restoration) {
     decode_restoration_mode(cm, rb);
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-restoration[y:%d,u:%d,v:%d]: off=%u\n",
+           cm->rst_info[0].frame_restoration_type,
+           cm->rst_info[1].frame_restoration_type,
+           cm->rst_info[2].frame_restoration_type,
+           rb->bit_offset);
+#endif
   }
   for (int plane = 0; plane < CCSO_NUM_COMPONENTS; plane++) {
     cm->ccso_info.ccso_enable[plane] = false;
   }
   if (!features->coded_lossless && seq_params->enable_ccso) {
     setup_ccso(cm, rb);
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-ccso[%d]: off=%u\n",
+           cm->ccso_info.ccso_frame_flag,
+           rb->bit_offset);
+#endif
   }
 
   features->tx_mode = read_tx_mode(rb, features->coded_lossless);
@@ -9821,6 +9982,16 @@ static int read_uncompressed_header(AV1Decoder *pbi,
 #else
       aom_rb_read_bit(rb);
 #endif  // CONFIG_REDUCED_TX_SET_EXT
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-modebits[tx:%d,refmode:%d,skipmode:%d,bawp:%d,warp:%d,redtxset:%d]: off=%u\n",
+           features->tx_mode,
+           current_frame->reference_mode,
+           current_frame->skip_mode_info.skip_mode_flag,
+           features->enable_bawp,
+           features->allow_warpmv_mode,
+           features->reduced_tx_set_used,
+           rb->bit_offset);
+#endif
 
   if (features->allow_ref_frame_mvs && !frame_might_allow_ref_frame_mvs(cm)) {
     aom_internal_error(&cm->error, AOM_CODEC_CORRUPT_FRAME,
@@ -9832,7 +10003,13 @@ static int read_uncompressed_header(AV1Decoder *pbi,
                        "Frame wrongly requests TIP mode");
   }
 
-  if (!frame_is_intra_only(cm)) read_global_motion(cm, rb);
+  if (!frame_is_intra_only(cm)) {
+    read_global_motion(cm, rb);
+#if DEBUG_FRAME_HDR
+    printf("HDR: post-gmv: off=%u\n",
+           rb->bit_offset);
+#endif
+  }
 
   cm->cur_frame->film_grain_params_present =
       seq_params->film_grain_params_present;
-- 
2.48.1

