From 561e2a8f1f652250ff4113be902e50fb5b99f0f8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20Storsj=C3=B6?= <martin@martin.st>
Date: Sat, 1 Apr 2023 01:43:39 +0300
Subject: [PATCH 23/23] WIP [libunwind] Fix get_proc_info for aarch64 SEH

---
 libunwind/src/libunwind.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libunwind/src/libunwind.cpp b/libunwind/src/libunwind.cpp
index 1bd18659b786..4ecce0487084 100644
--- a/libunwind/src/libunwind.cpp
+++ b/libunwind/src/libunwind.cpp
@@ -199,7 +199,7 @@ _LIBUNWIND_HIDDEN int __unw_get_proc_info(unw_cursor_t *cursor,
                        static_cast<void *>(cursor), static_cast<void *>(info));
   AbstractUnwindCursor *co = (AbstractUnwindCursor *)cursor;
   co->getInfo(info);
-  if (info->end_ip == 0)
+  if (info->start_ip == 0)
     return UNW_ENOINFO;
   return UNW_ESUCCESS;
 }
-- 
2.34.1

